rfc/terraform-provider.md

```diff
@@ -0,0 +1,92 @@
+# Constellation Terraform Provider
+
+The Constellation Terraform Provider allows its user to manage the full lifecycle of a Constellation cluster -- namely creation, updates, and deletion -- via Terraform.
+
+## Design Goals
+
+- The User needs to be able to perform Creation, Upgrades and Deletion of a Cluster through the provider.
+- Infrastructure provisioning should explicitly **not** be performed by the Constellation Terraform provider.
+- The User needs to receive some sort of Access Token to the Cluster (e.g. a Kubeconfig, whether long- or short-lived)
+when applying a configuration containing the provider and the resources listed below.
+
+## Terraform Configuration Layout for the Constellation Resource
+
+This resembles an examplary configuration of a Constellation cluster through Terraform. While the exact naming and struct layout
+of the individual components may change until the implementation, the structure of the overall configuration looks like this:
+
+```hcl
+terraform {
+  required_providers {
+    constellation = {
+      source  = "tbd/constellation"
+      version = "2.13.0"
+    }
+  }
+}
+
+provider "constellation" { }
+
+resource "constellation_cluster" "foo" {
+    uid = "bar"
+    name = "baz"
+    image = data.constellation_image.ref # or provide manually crafted values
+    kubernetes_version = "v1.27.6"
+    constellation_microservice_version = "lockstep" # or pinned to a specific version
+    debug = false
+    init_endpoint = "10.10.10.10" # should use public ip of LB resource, ideally also provisioned through TF
+    kubernetes_api_endpoint = "10.10.10.10" # should use public ip of LB resource, ideally also provisioned through TF
+    extra_microservices = {
+        csi_driver = true
+        # + more
+        # possiblly also constellation microservices with version and maybe service options,
+        # which would make constellation_microservice_version obsolete.
+        # exact API TBD
+    }
+    master_secret = "foo" # updating this would force recreation of the cluster
+    init_secret = "bar" # maybe derive from master_secret, updating this would force recreation of the cluster
+    network_config = {
+        # TBD
+        # should contain CIDRs for pod network, service cidr, node network... for Cilium
+        # and in-cluster Kubernetes API endpoint, e.g. for Kubelets
+    }
```

Comment:
Are users expected to fill this in, or can this be derived from the cloud provider name and the state of the cloud infrastructure?

Reply:
I don't know for sure yet, @3u13r may have more information here?

Reply:
We can get everything from the terraform as we also do today.

Reply:
But the values in this resource are the only thing that our provider gets to see, correct? Would we need to reference other Terraform resources, like

```
  network_config {
    flavor = "gcp"
    podCIDR = other_resource.vpc.cidr
    # ...
  }
```

Reply:
The latter, I think: https://github.com/edgelesssys/constellation/blob/f6c44ee5cc6f1d097c0b2096f6a8e289e2210730/terraform/infrastructure/gcp/outputs.tf#L35-L37

Reply:
We don't have to reference the other resource directly but there should be a direct link between the output of the vpc and the input of the Constellation provider (as Markus mentioned).

Reply:
Is this something that needs to be mentioned in the RFC? IMO, it should be agnostic about where the value comes from (output of another resource or directly supplied)

Reply:
Only if the goal is to replace the TBD with an explicit list, which I don't think is needed. Maybe we could mention that some fields here need to be populated by other resources' outputs, but maybe that is obvious.